keps/sig-api-machinery/20190425-structural-openapi.md

```diff
@@ -0,0 +1,812 @@
+
+We want that specified types and object field names are mechanically extractable with an understanding of `properties`, `items`, `additionaProperties` and `type` only. Other constructs like logical junctors should be usable to express complex validations, but should not be essential for the specification of the type structure of JSON values.
+
+We reach this goal in defining _structural schema_ by following this guiding principle:
+
+**Structural Completeness**: all types and the possible object fields within a JSON object must be apparent from a structural schema without understanding the logical junctors `anyOf`, `allOf`, `oneOf` and `not`.
+
+### Formal Definition
+
+We define a _structural schema_ by choosing another data structure representation of [`JSONSchemaProps`](https://github.com/kubernetes/kubernetes/blob/d0c3b708026f681f7467d36dcf9696f89ccfcfab/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types_jsonschema.go#L20), while adding vendor extension fields for:
+* `x-kubernetes-preserve-unknown-fields: true` – recursive version of `additionalProperties: true`, influencing pruning to not prune unknown fields (compare [pruning KEP](https://github.com/kubernetes/enhancements/pull/709) for more details).
+* `x-kubernetes-embedded-resource: true` – marks a RawExtension field to embed a complete `runtime.Object` like object with `TypeMeta` and `ObjectMeta`.
+* `x-kubernetes-unions: [...]` as defined in [KEP: Union Types](https://github.com/kubernetes/enhancements/pull/926)
+* `x-kubernetes-int-or-string: true` – marks a `IntOrString` polymorphic type.
+
+Observe that any validated `JSONSchemaProps` can be represented as `StructuralSchema`. Validation allows us to drop some unsupported fields (which are part of `JSONSchemaProps` but which we forbid in CRDs):
+
+```go
+type StructuralSchema struct {
+	Items                *StructuralSchema
+	Properties           map[string]StructuralSchema
+
+	Generic
+	Extensions
+	
+	*ValueValidation
+}
+
+type Extensions struct {
+	// x-kubernetes-unions lists unions, i.e. sets of fields which are 
+	// mutual exclusive.
+	XUnions []Union
+	
+        // x-kubernetes-preserve-unknown-fields stops the API server 
+        // decoding step from pruning fields which are not specified 
+        // in the validation schema. This affects fields recursively,
+        // but switches back to normal pruning behaviour if nested 
+        // properties or additionalProperties are specified in the schema.
+        XPreserveUnknownFields bool
+        
+        // x-kubernetes-embedded-resource defines that the value is an 
+        // embedded Kubernetes runtime.Object, with TypeMeta and 
+        // ObjectMeta. The type must be object. It is allowed to further
+        // restrict the embedded object. Both ObjectMeta and TypeMeta
+        // are validated automatically. x-kubernetes-preserve-unknown-fields
+        // must be true.
+        XEmbeddedResource bool
```

Comment:
do we want to consider https://github.com/kubernetes/kubernetes/issues/76965 here, to allow recording one or more GVKs expected to be embedded here?

Reply:
I tend to postpone this whole `$ref` topic to later and to another KEP. Do you see that we have to decide anything now? We can add a `x-kubernetes-embedded-resource-gvk` is we want later.

Reply:
sounds good